Other than the minor quibbles about the content of the message, this seems ok, but it seems like it just papers over the issue, which is deeper than the test code, given that the test just calls a public method (`ADB.close()`).

Also, I see a similar exception when running only the "main" test script, whether or not libadb is open in another process:

```
08-05 15:50 > make test TEST=main
…
info: r2d2b2g: Closing ADB
info: r2d2b2g: mac64/B2G.app/Contents/MacOS/b2g terminated with 15
info: r2d2b2g: [JavaScript Error: "TypeError: socket is undefined" {file: "resource://gre/modules/XPIProvider.jsm -> file:///var/folders/lp/8t_7y24119720_hjp5wc_4cr0000gn/T/tmp6kbnDY.mozrunner/extensions/r2d2b2g@mozilla.org/bootstrap.js -> resource://gre/modules/commonjs/toolkit/loader.js -> resource://r2d2b2g-at-mozilla-dot-org/r2d2b2g/lib/adb/adb-device-tracker.js" line: 115}]
info: r2d2b2g: Simulator.onMessage toggle
!!! error running onStopped callback: TypeError: callback is not a function
Traceback (most recent call last):
  File "/Users/myk/Mozilla/r2d2b2g/addon-sdk/bin/cfx", line 33, in <module>
    cuddlefish.run()
  File "/Users/myk/Mozilla/r2d2b2g/addon-sdk/python-lib/cuddlefish/__init__.py", line 928, in run
    bundle_sdk=options.bundle_sdk)
  File "/Users/myk/Mozilla/r2d2b2g/addon-sdk/python-lib/cuddlefish/runner.py", line 743, in run_app
    OUTPUT_TIMEOUT, test_name, parseable)
Exception: Test output exceeded timeout (60s).
make: *** [test] Error 1
```

But perhaps this is a different issue?

--- a/addon/lib/adb/adb-running-checker.js
+++ b/addon/lib/adb/adb-running-checker.js
@@ -40,7 +40,7 @@ exports.check = function check() {
         let { length, data } = client.unpackPacket(aData);
         debug("length: ", length, "data: ", data);
         socket.close();
-        deferred.resolve(true);
+        deferred.resolve(data.indexOf("001f") != -1);
         break;
       default:
         debug("Unexpected State: " + state);

--- a/addon/lib/adb/adb.js
+++ b/addon/lib/adb/adb.js
@@ -268,7 +268,10 @@ exports._startAdbInBackground = function startAdbInBackground() {
 
   serverWorker.emit("init", { libPath: libPath }, function initack() {
     serverWorker.emit("start", { port: 5037, log_path: File.join(TmpD, "adb.log") }, function started(res) {
-      console.debug("adb server thread returned: " + res.result);
+      console.debug("adb server thread returned: " + res.ret);
+      if (res.ret == -1) {
+        Services.obs.notifyObservers(null, "adb-port-in-use", null);
+      }
     });
   });
 

--- a/addon/test/test-main.js
+++ b/addon/test/test-main.js
@@ -4,7 +4,7 @@
 
 /* Note: all tests that depend on firefox state should be put in this
  * file.  Re-requiring main in other tests causes the simulator to be
- * reinstantiated. 
+ * reinstantiated.
  */
 
 //==========================
@@ -25,6 +25,7 @@ exports = (function(exports) {
   Cu.import("resource://gre/modules/Services.jsm");
 
   let isPhonePluggedIn = null;
+  let isPortInUse = false;
 
   let observer = {
     QueryInterface: XPCOMUtils.generateQI([Ci.nsIObserver,
@@ -40,6 +41,9 @@ exports = (function(exports) {
           break;
         case "adb-device-disconnected":
           isPhonePluggedIn = false;
+          break;
+        case "adb-port-in-use":
+          isPortInUse = true;
           break;
       }
     }
@@ -58,6 +62,8 @@ exports = (function(exports) {
         }
         Services.obs.addObserver(observer, "adb-device-connected", true);
         Services.obs.addObserver(observer, "adb-device-disconnected", true);
+        Services.obs.addObserver(observer, "adb-ready", true);
+        Services.obs.addObserver(observer, "adb-port-in-use", true);
         assert.pass("Started");
         done();
       });
@@ -93,6 +99,14 @@ exports = (function(exports) {
   exports["test ab list devices"] = function(assert, done) {
     // Give adb 2 seconds to startup
     Timer.setTimeout(function listDevices() {
+      if (isPortInUse) {
+        isPhonePluggedIn = false;
+        assert.fail("Error: Port 5037 is in use.\nHave you opened the " +
+                    "Simulator in a different Firefox profile?\nMake " +
+                    "sure you quit that process before running the tests.");
+        done();
+        return;
+      }
       ADB.listDevices().then(
         function success(e) {
           if (ADB.didRunInitially && e[0]) {
@@ -241,7 +255,13 @@ exports = (function(exports) {
 
   exports["test zz after"] = function(assert, done) {
     if (ADB.didRunInitially) {
-      ADB.close();
+      try {
+        ADB.close();
+      } catch (e) {
+        if (!isPortInUse) {
+          throw e;
+        }
+      }
     }
     assert.pass("Done!");
     done();

--- a/android-tools/adb-bin/adb.cpp
+++ b/android-tools/adb-bin/adb.cpp
@@ -1192,7 +1192,7 @@ void build_local_name(char* target_str, size_t target_size, int server_port)
   snprintf(target_str, target_size, "tcp:%d", server_port);
 }
 
-void * server_thread(void * args) {
+int server_thread(void * args) {
   adb_sysdeps_init();
 
   struct adb_main_input* input = (struct adb_main_input*)args;
@@ -1247,8 +1247,7 @@ void * server_thread(void * args) {
     build_local_name(local_name, sizeof(local_name), server_port);
     if(install_listener(local_name, "*smartsocket*", NULL, 0)) {
         D("Error installing listener\n");
-        return NULL;
-
+        return -1;
     }
 
     if (is_daemon)
@@ -1481,11 +1480,8 @@ int handle_host_request(char *service, transport_type ttype, char* serial, int r
   //
     // returns our value for ADB_SERVER_VERSION
     if (!strcmp(service, "version")) {
-        char version[12];
-        snprintf(version, sizeof version, "%04x", ADB_SERVER_VERSION);
-        snprintf(buf, sizeof buf, "OKAY%04x%s", (unsigned)strlen(version), version);
-        writex(reply_fd, buf, strlen(buf));
-        return 0;
+        printf("the version service is disabled.\n");
+        return -1;
     }
 
     if(!strncmp(service,"get-serialno",strlen("get-serialno"))) {

--- a/android-tools/adb-bin/exports.cpp
+++ b/android-tools/adb-bin/exports.cpp
@@ -116,8 +116,7 @@ DLL_EXPORT void on_kill_io_pump(atransport * t, bool (*close_handle_func)(ADBAPI
   // NOTE: input_args is free'd with `free` so must be alloc'd with malloc.
   //       This call loops forever.
   DLL_EXPORT int main_server(struct adb_main_input * input_args) {
-    server_thread((void *)input_args);
-    return 0;
+    return server_thread((void *)input_args);
   }
 
 #ifdef __APPLE__

--- a/android-tools/adb-bin/threads.h
+++ b/android-tools/adb-bin/threads.h
@@ -3,7 +3,7 @@
 
 #include "adb.h"
 
-void * server_thread(void * args);
+int server_thread(void * args);
 #ifdef __APPLE__
 void * RunLoopThread(void * unused);
 #else

